### PR TITLE
Add loading of pre and post modules

### PR DESCRIPTION
**Pull request by lumurillo**

### CODE REVIEW

File: examples/external/Makefile [hunk 1]
```diff
@@ -10,8 +10,8 @@
 
 R2IVERSION=0.0
 
-CXXFLAGS:= -std=c++11 $(shell pkg-config --cflags r2inference-$(R2IVERSION)) -pthread
-LIBS:=$(shell pkg-config --libs r2inference-$(R2IVERSION)) -pthread
+CXXFLAGS:= -std=c++11 $(shell pkg-config --cflags r2inference-$(R2IVERSION)) -pthread $(shell pkg-config --cflags glib-2.0) $(shell pkg-config --cflags gmodule-2.0) 
+LIBS:=$(shell pkg-config --libs r2inference-$(R2IVERSION)) -pthread $(shell pkg-config --libs glib-2.0) $(shell pkg-config --libs gmodule-2.0)
```
Comment (lumurillo) [inline examples/external/Makefile:14]: I had to add these flags to make this example to compile correctly, but shouldn't these flags come from R2Inference package config file?

File: meson.build [hunk 2]
```diff
@@ -106,7 +110,7 @@ if get_option('enable-onnxrt-acl')
     arm_compute_core_dep = declare_dependency(dependencies: arm_compute_core)
     arm_compute_graph = cpp.find_library('arm_compute_graph', required: true)
     arm_compute_graph_dep = declare_dependency(dependencies: arm_compute_graph)
-    lib_onnxrt_acl_dep = [arm_compute_dep, arm_compute_core_dep, arm_compute_graph_dep, lib_onnxrt_dep]
+    lib_onnxrt_acl_dep = [arm_compute_dep, arm_compute_core_dep, arm_compute_graph_dep, lib_onnxrt_dep, glib, gmodule]
```
Comment (jafet-chaves) [inline meson.build:113]: ONNXRT OpenVINO will probably need these dependencies aswell

File: r2i/iloader.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/* Copyright (C) 2018 RidgeRun, LLC (http://www.ridgerun.com)
```
Comment (jafet-chaves) [inline r2i/iloader.cc:1]: 2020. Check every other modified file for this.

File: r2i/ipreprocessing.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  virtual std::vector<std::vector<int>> getAvailableDataSizes() = 0;
+
+  /**
+   * \brief Default destructor
+   */
+  virtual ~IPreprocessing () {};
+};
+
+
+}
```
Comment (jafet-chaves) [inline r2i/ipreprocessing.h:58]: Add //  r2i. Check for this in other modified files.

File: r2i/ipreprocessing.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+extern "C" {
```
Comment (jafet-chaves) [inline r2i/ipreprocessing.h:60]: Why we need this?

Comment (lumurillo) [reply r2i/ipreprocessing.h:60]: This is needed to load C symbols during the loading of the pre and post modules. The gmodule functions work better with C symbols.

File: r2i/iloader.cc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  PreprocessFactoryFunc factory;
+
+  GModule *module = LoadModule(in_path.c_str(), error);
+  if (RuntimeError::EOK != error.GetCode()) {
+    return nullptr;
+  }
+
+  if (!g_module_symbol (module, MODULE_FACTORY_PREPROCESSING_SYMBOL,
+                        (gpointer *) & factory)) {
+    error.Set(RuntimeError::UNKNOWN_ERROR, g_module_error ());
```
Comment (michaelgruner) [inline r2i/iloader.cc:35]: Add a new error code.

File: r2i/iloader.cc [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    return nullptr;
+  }
+
+  IPreprocessing *preprocessing = (IPreprocessing *) factory();
+  if (!preprocessing) {
+    error.Set(RuntimeError::UNKNOWN_ERROR, "Error casting preprocessing object.");
```
Comment (michaelgruner) [inline r2i/iloader.cc:41]: This is not a casting error.

File: r2i/ipostprocessing.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#ifndef R2I_IPOSTPROCESSING_H
+#define R2I_IPOSTPROCESSING_H
+
+#include <r2i/runtimeerror.h>
```
Comment (rrcarlosrodriguez) [inline r2i/ipostprocessing.h:15]: alphabetical order

File: r2i/Makefile.am [hunk 9]
```diff
@@ -60,6 +60,7 @@ r2iincludedir = @R2IINCLUDEDIR@/r2i
 libr2inference_@RR_PACKAGE_VERSION@_la_SOURCES =    \
         datatype.cc                                 \
         iframeworkfactory.cc                        \
+        iloader.cc                                  \
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply r2i/Makefile.am:63]: Ok, I agree. I am going to add a base class that implements this interface.

File: meson.build [hunk 10]
```diff
@@ -27,11 +27,15 @@ cpp_args = ['-DHAVE_CONFIG_H']
...
+# define common dependencies
+glib = dependency('glib-2.0')
+gmodule = dependency('gmodule-2.0')
+
 # Define library dependencies for Tensorflow support
 if get_option('enable-tensorflow')
   tensorflow = cpp.find_library('tensorflow', required: true)
   tensorflow_dep = declare_dependency(dependencies: tensorflow)
-  lib_tf_dep = [tensorflow_dep]
+  lib_tf_dep = [tensorflow_dep, glib, gmodule]
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply meson.build]: These are common dependencies between all backends. I had refactored the way to add these common dependencies.

File: tests/unit/r2i/loader.cc [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  }
+};
+
+TEST (Loader, LoadPreprocessingWithNullPath) {
+  std::string string_path;
+
+  std::shared_ptr<r2i::IPreprocessing> preprocessing = loader.LoadPreprocessing(
+        string_path, error);
+
+  CHECK (r2i::RuntimeError::EOK != error.GetCode());
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply tests/unit/r2i/loader.cc]: Agree. I just changed it.

File: examples/external/Makefile [hunk 12]
```diff
@@ -10,7 +10,7 @@
 
 R2IVERSION=0.0
 
-CXXFLAGS:= -std=c++11 $(shell pkg-config --cflags r2inference-$(R2IVERSION)) -pthread
+CXXFLAGS:= -std=c++11 $(shell pkg-config --cflags r2inference-$(R2IVERSION)) -pthread $(shell pkg-config --cflags glib-2.0 gmodule-2.0)
```
Comment (jafet-chaves) [inline examples/external/Makefile:13]: I think you can do:
CXXFLAGS:= -std=c++11 $(shell pkg-config --cflags glib-2.0 gmodule-2.0 r2inference-$(R2IVERSION)) -pthread

File: r2i/loader.cc [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+std::shared_ptr<r2i::IModel> Loader::Load (const std::string & /*in_path*/,
+    r2i::RuntimeError &error) {
+  error.Set(RuntimeError::NOT_IMPLEMENTED, "Loader::Load method not implemented");
+  return nullptr;
+}
+
+std::shared_ptr<IPreprocessing> Loader::LoadPreprocessing(
+  const std::string &in_path, RuntimeError &error) {
+  PreprocessFactoryFunc factory;
```
Comment (mmontero) [inline r2i/loader.cc:32]: initialize all the variables

File: r2i/loader.cc [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+#define MODULE_FACTORY_POSTPROCESSING_SYMBOL "factory_make_postprocessing"
+
+typedef
+r2i::IPreprocessing *(*PreprocessFactoryFunc) (void);
+typedef
+r2i::IPostprocessing *(*PostprocessFactoryFunc) (void);
+
+namespace r2i {
+
+std::shared_ptr<r2i::IModel> Loader::Load (const std::string & /*in_path*/,
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply r2i/loader.cc:24]: This is to avoid the error caused by an unused variable.

File: r2i/loader.h [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+   * \param in_path A string with the path to the dynamic library.
+   * \param error [out] RuntimeError with a description of an error.
+   * \return Loaded module or Null if the loading failed.
+   */
+  GModule *LoadModule(const gchar *in_path, RuntimeError &error);
+};
+
+}
+
+#endif // R2I_ILOADER_H
```
Comment (mmontero) [inline r2i/loader.h:52]: Add new line at the end of file

**Pull request closed without merging**